torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,159 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// kRpcRetryDuration is the initial duration between consecutive RPC retries.
+constexpr auto kRpcRetryDuration = std::chrono::milliseconds(1000);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+constexpr float kBackoff = 1.5;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  rpcAgentRunning_.store(true);
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  if (rpcAgentRunning_.load()) {
+    shutdown();
+  }
+}
+
+void RpcAgent::shutdown() {
+  if (!rpcAgentRunning_.exchange(false)) {
+    return;
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+  shutdownImpl();
+}
+
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  TORCH_CHECK(maxRetries >= 0, "maxRetries cannot be negative.");
+  auto originalFuture = std::make_shared<FutureMessage>();
+  steady_clock_time_point newTime = computeNewRpcRetryTime(/* retryCount */ 0);
+  Message& msg = message;
+  auto fm = send(to, std::move(message));
+  auto firstRetryRpc = std::make_shared<RpcRetryInfo>(
+      to, std::move(msg), originalFuture, /* retryCount */ 0, maxRetries);
```

Comment:
`msg` here is a reference to a **moved** `Message`? The move in `send` should have already cleared the payload in `message`. Is this intentional?

Reply:
btw, I think this line ends up being correct if line 49 is "Message msg = message;" rather than "Message& msg = message;"

ideally add a short comment, and use variable names, to make the copy explicit, to avoid future regressions, e.g.

    Message msgCopy = message; // make a copy of the original for retries
    auto fm = send(to, std::move(message));
   auto firstRetryRpc = std::make_shared<RpcRetryInfo>(
      to, std::move(msgCopy), originalFuture, /* retryCount */ 0, maxRetries);